pkg/util/strategicpatch/patch.go

```
@@ -0,0 +1,454 @@
+	}
+
+	// First look for any special __patch elements.
+	patchWithoutSpecialElements := []interface{}{}
+	strict := false
+	for _, v := range patch {
+		typedV := v.(map[string]interface{})
+		patchType, ok := typedV[specialKey]
+		if ok {
+			if patchType == "delete" {
+				mergeValue, ok := typedV[mergeKey]
+				if ok {
+					_, originalKey, found := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
+					if found {
+						// Delete the element at originalKey.
+						original = append(original[:originalKey], original[originalKey+1:]...)
+					}
+				} else {
+					return nil, fmt.Errorf("delete patch type with no merge key defined")
+				}
+			} else if patchType == "strict" {
+				strict = true
+				// Continue iterating through the array to prune any other __patch elements.
+			} else {
+				return nil, fmt.Errorf("unknown patch type found: %s", patchType)
+			}
+		} else {
+			patchWithoutSpecialElements = append(patchWithoutSpecialElements, v)
+		}
+	}
+
+	if strict {
+		return patchWithoutSpecialElements, nil
+	}
+
+	patch = patchWithoutSpecialElements
+
+	// Merge patch into original.
+	for _, v := range patch {
+		// Because earlier we confirmed that all the elements are maps.
+		typedV := v.(map[string]interface{})
+		mergeValue, ok := typedV[mergeKey]
+		if !ok {
+			return nil, fmt.Errorf("all list elements need the merge key %s", mergeKey)
+		}
+		// If we find a value with this merge key value in original, merge the
+		// maps. Otherwise append onto original.
+		originalMap, originalKey, found := findMapInSliceBasedOnKeyValue(original, mergeKey, mergeValue)
+		if found {
+			var mergedMaps interface{}
+			var err error
+			// Merge into original.
+			mergedMaps, err = mergeMap(originalMap, typedV, elemType)
+			if err != nil {
+				return nil, err
+			}
+			original[originalKey] = mergedMaps
+		} else {
+			original = append(original, v)
+		}
+	}
+
+	return original, nil
+}
+
+// This panics if any element of the slice is not a map.
```

Why panic rather than return an error?